### ISSUE TITLE
Inappropriate safe browsing implementation

### ISSUE THREAD

**Issue by ZeroClover** [AdguardTeam/AdGuardHome#2857]
### Prerequisites

- [x] I am running the latest version
- [x] I checked the documentation and found no answer
- [x] I checked to make sure that this issue has not already been filed

### Issue Details

* **Version of AdGuard Home server:**
v0.105.2
* **How did you install AdGuard Home:**
Snapcraft
* **How did you setup DNS configuration:**
Dedicated server
* **CPU architecture:**
AMD64
* **Operating system and version:**
Debian 10

### Expected Behavior
AdGuard Home uses Google Safe Browsing Update API

### Actual Behavior
AdGuard Home Send **ALL** my queries to my AdGuard's "Home Protection" DNS server

### Additional Information

After reading the AGH code I was shocked to discover that the way AGH implements safe browsing is by sending queries to AdGuard's official "Family Protected" DNS server via DoH.

I believe this implementation is inappropriate due to the following factors.

a. The authorities in mainland China have blocked well-known encrypted DNS providers, including AdGuard DNS, so it is almost impossible for users in mainland China to use AGH safe browsing feature properly.

b. The Google Safe Browsing Update API allows clients to perform initial checks locally by enabling them to download a list of safe browsing hashes in advance, and only if they match will they need to send further requests to the server to perform the checks. This not only makes processing much faster, but also improves privacy. The server does not know all of the client's browsing history.

This issue was discovered when I received feedback from users that several of my AdGuard Home servers deployed in mainland China were experiencing serious performance issues. I checked the logs and found that AdGuard was taking more than 3000ms to process DNS queries.

<details><summary>Screenshot:</summary>

![image](https://user-images.githubusercontent.com/13190004/111920021-a669ab00-8ac7-11eb-8794-6493b011c080.png)

</details>

After reviewing the debug logs, I found that AdGuard Home was trying to send DNS requests to https://dns-family.adguard.com/dns-query, due to AdGuard DNS is fully blocked in mainland China, so it is impossible for servers in mainland China to access AdGuard DNS.

PS: Google Safe Browsing is available in mainland China, it is not blocked by the mainland Chinese government, and it also has servers in mainland China.

In addition, even if you do not use Google-related services for privacy reasons, you should consider other implementations than sending a request at every query.

**Comment by Macqael** [comment]
@ameshkov Other easy way to deal with this very hight latency (for waiting a more heavy (in dev time) solution) it's when you enable safe browsing or family you test with "test-safe.adguard.org" or "test-familly.adguard.org" or simply "test.adguard.org" and he must respond a predetermined answer like "bad domain" if not the safe browsing isn't enabled.

**Comment by Aikatsui** [comment]
Duplicate of #1834

>a. The authorities in mainland China have blocked well-known encrypted DNS providers, including AdGuard DNS, so it is almost impossible for mainland China users to use AGH safe browsing feature properly.

This is not something only belong to china and some other countries block as well. You can check to set-up adguard DNS IP addresses in adguardhome.yaml and see if it works.

>slow access to AdGuard DNS can significantly impact DNS query performance.
> Since AdGuard DNS's Anycast only covers a small portion of the world (relative to Google), turning on Safe Browsing in other areas will cause significant DNS query performance degradation.

Large enterprises always offer better connectivity and minimize performance issues in multiple ways. It takes long time for adguard to become google ;)

>At a minimum, alert the Dashboard when the safe browsing query's performance is too low

Otherhand you can use local providers and it'll be the best solution for avoid such problems. Past i added your country providers to KB article, some are also have explained about protect users privacy and the technology is same

https://kb.adguard.com/en/general/dns-providers#ali-dns (it comes with protection)
https://kb.adguard.com/en/general/dns-providers#dnspod-public-dns (also comes with protection)
https://kb.adguard.com/en/general/dns-providers#114dns (normal, safe and family)
https://kb.adguard.com/en/general/dns-providers#onedns (pure and block)

**Comment by ZeroClover** [comment]
@Aikatsui 

> This is not something only belong to china and some other countries block as well. You can check to set-up adguard DNS IP addresses in adguardhome.yaml and see if it works but i guess they block providers entirely regardless encrypted.

I think the GFW blocking is based on SNI, so it is impossible to access AdGuard DNS in mainland China by any method.

> Large enterprises always offer better connectivity and minimize performance issues in multiple ways. It takes long time for adguard to become google ;)

Yes, I know, and I appreciate the effort AdGuard has put in.

> Otherhand you can use local providers and it'll be the best solution for avoid such problems. Past i added your country providers to KB article, some are also have explained about protect users privacy and the technology is same

I use my own DNS as an upstream, and I use an international Ethernet private line to access recursive DNS outside of China, but the AdGuard Family DNS for safe browsing is hardcoded in dnsfilter/safebrowsing.go, and other DNS should not support AdGuard's implementation.

I've solved this problem with my own approach so far. But using an implementation similar to Google Safe Browsing Update API can still help users around the world reduce the extra latency caused by enabling Safe Browsing.

**Comment by DandelionSprout** [comment]
The bigger problem at hand here, is that AdGuard's anti-malware and parental-protection modules are too closed-source even for me: No one knows their contents, despite them seemingly just being server-side filterlists.

Had the modules' list of blocked domains gone public, they could be deployed admin-side in AdGuard Home as a filterlist, bypassing AdGuard DNS' servers entirely and avoiding connection issues.

**Comment by ameshkov** [comment]
Do you mean that the lists are closed? Because the source code for working with them is open, it's a part of AGH.

**Comment by ameshkov** [comment]
> Had the modules' lists of blocked domains gone public

It's not that easy. There're sources of this data that we pay for, we can't just make this all open and available to everyone as this way we'll basically redistribute paid stuff.

**Comment by Macqael** [comment]
> > Had the modules' lists of blocked domains gone public
> 
> It's not that easy. There're sources of this data that we pay for, we can't just make this all open and available to everyone as this way we'll basically redistribute paid stuff.

Can you at least do like quad9 tell what are the name of your sources (and add it in the doc)

**Comment by ameshkov** [comment]
Hm, probably, we'll need to check with the sources.

**Comment by ammnt** [comment]
Yeah, this function is one of the reasons for me to use AGH, but its terrible performance is disappointing😪

**Comment by Macqael** [comment]
> Yeah, this function is one of the reasons for me to use AGH, but its terrible performance is disappointing😪

in my case on server who have poor performances with these feature I increase the duration/size of the cache of these two API.
It help a little bit another option is to disable them and use directly ADG DNS as upstream.

**Comment by ammnt** [comment]
@Macqael, did you mean `safebrowsing_cache_size`? What specific value did you set? Thank you.

**Comment by ammnt** [comment]
DandelionSprout is actually right👆It makes sense to publush lists of adult and phishing sites in a separate repo, just like AdGuard DNS filter. So we can use them separately and help you with false positives🙏

**Comment by ammnt** [comment]
@ainar-g & @ameshkov, today is the fifth day I deployed AdGuard Home on a server with an NVMe drive and there is no longer a problem with low performance of antiphishing module. Can this be related?

**Comment by Quijib0** [comment]
I was/am having this same issue.  Originally, I had thought it was my route from my ISP.  Turns out, the AdG server got added to my subscribed BanIP DOH server list.  I really don't wish to use DOH on my home network due to the back door that it creates for malware.  So, I won't be using the browsing security feature.

  Anyway, I verified that removing the ban brought the resolving performance back to normal with browsing security enabled.

**Comment by timkgh** [comment]
Seeing the same performance issues here. Enabling browsing security and parental control adds about 40ms of overhead, even when resolving local names on the LAN (so these lookups went from <1ms to >40ms).
This should be documented and made very clear how enabling the 2 options work and the performance impact.
I thought that these lists would be downloaded locally, but it appears that the requests are sent to the AdGuard DNS servers. Then what's the point using other fast upstreams, might as well use AdGuard DNS directly.

**Comment by zhqu1148980644** [comment]
These features are optional; the AGH dns server should work properly regardless of the reasons.
